File: src/cli/commands/list.rs
use clap::Args;
use serde::Serialize;

use crate::container::{ContainerRuntime, ContainerRuntimeFactory};
use crate::error::Result;
use crate::labels;

/// List running MCP servers
#[derive(Args, Debug)]
pub struct ListCommand {
    /// Show all containers (default shows just running)
    #[arg(short, long)]
    pub all: bool,

    /// Output in JSON format
    #[arg(short, long)]
    pub json: bool,
}

/// Container information for JSON output
#[derive(Serialize)]
struct ContainerOutput {
    id: String,
    name: String,
    image: String,
    state: String,
    transport: String,
}

impl ListCommand {
    /// Run the command
    pub async fn execute(&self) -> Result<()> {
        // Create container runtime
        let runtime = ContainerRuntimeFactory::create().await?;
        
        // Execute with the runtime
        self.execute_with_runtime(runtime).await
    }
    
    /// Run the command with a specific runtime (for testing)
    pub async fn execute_with_runtime(&self, runtime: Box<dyn ContainerRuntime>) -> Result<()> {
        // List containers
        let containers = runtime.list_containers().await?;

        // Filter containers if not showing all
        let containers = if !self.all {
            containers
                .into_iter()
                .filter(|c| c.state.contains("running"))
                .collect()
        } else {
            containers
        };

        if self.json {
            self.print_json_output(&containers)?;
        } else {
            self.print_text_output(&containers);
        }

        Ok(())
    }

    /// Print container information in JSON format
    fn print_json_output(&self, containers: &[crate::container::ContainerInfo]) -> Result<()> {
        let output: Vec<ContainerOutput> = containers.iter().map(|container| {
            // Truncate container ID to first 12 characters (similar to Docker)
            let truncated_id = container.id.chars().take(12).collect::<String>();
            
            // Get transport from labels
            let transport = labels::get_transport(&container.labels);
            
            ContainerOutput {
                id: truncated_id,
                name: container.name.clone(),
                image: container.image.clone(),
                state: container.state.clone(),
                transport: transport.to_string(),
            }
        }).collect();
        
        println!("{}", serde_json::to_string_pretty(&output)?);
        Ok(())
    }

    /// Print container information in text format
    fn print_text_output(&self, containers: &[crate::container::ContainerInfo]) {
        println!("{:<12} {:<20} {:<40} {:<15} {:<10}", "CONTAINER ID", "NAME", "IMAGE", "STATE", "TRANSPORT");
        for container in containers {
            // Truncate container ID to first 12 characters (similar to Docker)
            let truncated_id = container.id.chars().take(12).collect::<String>();
            
            // Get transport from labels
            let transport = labels::get_transport(&container.labels);
            
            println!(
                "{:<12} {:<20} {:<40} {:<15} {:<10}",
                truncated_id, container.name, container.image, container.state, transport
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cli::commands::tests::{create_mock_runtime, create_test_container_info};
    
    #[tokio::test]
    async fn test_list_command_all_containers() -> Result<()> {
        // Create a mock runtime
        let mut mock_runtime = create_mock_runtime();
        
        // Set up expectations for list_containers
        mock_runtime.expect_list_containers()
            .times(1)
            .returning(|| {
                Ok(vec![
                    create_test_container_info("container1", "test-container-1", "Up 10 minutes"),
                    create_test_container_info("container2", "test-container-2", "Exited (0) 5 minutes ago"),
                ])
            });
        
        // Create and execute the command
        let cmd = ListCommand { all: true, json: false };
        let result = cmd.execute_with_runtime(Box::new(mock_runtime)).await;
        
        // Verify the result
        assert!(result.is_ok());
        
        Ok(())
    }
    
    #[tokio::test]
    async fn test_list_command_running_containers() -> Result<()> {
        // Create a mock runtime
        let mut mock_runtime = create_mock_runtime();
        
        // Set up expectations for list_containers
        mock_runtime.expect_list_containers()
            .times(1)
            .returning(|| {
                Ok(vec![
                    create_test_container_info("container1", "test-container-1", "Up 10 minutes"),
                    create_test_container_info("container2", "test-container-2", "Exited (0) 5 minutes ago"),
                ])
            });
        
        // Create and execute the command
        let cmd = ListCommand { all: false, json: false };
        let result = cmd.execute_with_runtime(Box::new(mock_runtime)).await;
        
        // Verify the result
        assert!(result.is_ok());
        
        Ok(())
    }

    #[tokio::test]
    async fn test_list_command_json_output() -> Result<()> {
        // Create a mock runtime
        let mut mock_runtime = create_mock_runtime();
        
        // Set up expectations for list_containers
        mock_runtime.expect_list_containers()
            .times(1)
            .returning(|| {
                Ok(vec![
                    create_test_container_info("container1", "test-container-1", "Up 10 minutes"),
                ])
            });
        
        // Create and execute the command with JSON output
        let cmd = ListCommand { all: true, json: true };
        let result = cmd.execute_with_runtime(Box::new(mock_runtime)).await;
        
        // Verify the result
        assert!(result.is_ok());
        
        Ok(())
    }
}

File: cmd/vibetool/list.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"text/tabwriter"

	"github.com/spf13/cobra"

	"github.com/stacklok/vibetool/pkg/container"
	"github.com/stacklok/vibetool/pkg/labels"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List running MCP servers",
	Long:  `List all MCP servers managed by Vibe Tool, including their status and configuration.`,
	RunE:  listCmdFunc,
}

var (
	listAll    bool
	listFormat string
)

// ContainerOutput represents container information for JSON output
type ContainerOutput struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Image     string `json:"image"`
	State     string `json:"state"`
	Transport string `json:"transport"`
	Port      int    `json:"port"`
}

func init() {
	listCmd.Flags().BoolVarP(&listAll, "all", "a", false, "Show all containers (default shows just running)")
	listCmd.Flags().StringVar(&listFormat, "format", "text", "Output format (json or text)")
}

func listCmdFunc(cmd *cobra.Command, args []string) error {
	// Create context
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Create container runtime
	runtime, err := container.NewFactory().Create(ctx)
	if err != nil {
		return fmt.Errorf("failed to create container runtime: %v", err)
	}

	// List containers
	containers, err := runtime.ListContainers(ctx)
	if err != nil {
		return fmt.Errorf("failed to list containers: %v", err)
	}

	// Filter containers to only show those managed by Vibe Tool
	var vibeToolContainers []container.ContainerInfo
	for _, c := range containers {
		if labels.IsVibeToolContainer(c.Labels) {
			vibeToolContainers = append(vibeToolContainers, c)
		}
	}

	// Filter containers if not showing all
	if !listAll {
		var runningContainers []container.ContainerInfo
		for _, c := range vibeToolContainers {
			if c.State == "running" {
				runningContainers = append(runningContainers, c)
			}
		}
		vibeToolContainers = runningContainers
	}

	if len(vibeToolContainers) == 0 {
		fmt.Println("No MCP servers found")
		return nil
	}

	// Output based on format
	switch listFormat {
	case "json":
		return printJSONOutput(vibeToolContainers)
	default:
		printTextOutput(vibeToolContainers)
		return nil
	}
}

// printJSONOutput prints container information in JSON format
func printJSONOutput(containers []container.ContainerInfo) error {
	var output []ContainerOutput

	for _, c := range containers {
		// Truncate container ID to first 12 characters (similar to Docker)
		truncatedID := c.ID
		if len(truncatedID) > 12 {
			truncatedID = truncatedID[:12]
		}

		// Get container name from labels
		name := labels.GetContainerName(c.Labels)
		if name == "" {
			name = c.Name // Fallback to container name
		}

		// Get transport type from labels
		transport := labels.GetTransportType(c.Labels)
		if transport == "" {
			transport = "unknown"
		}

		// Get port from labels
		port, err := labels.GetPort(c.Labels)
		if err != nil {
			port = 0
		}

		output = append(output, ContainerOutput{
			ID:        truncatedID,
			Name:      name,
			Image:     c.Image,
			State:     c.State,
			Transport: transport,
			Port:      port,
		})
	}

	// Marshal to JSON
	jsonData, err := json.MarshalIndent(output, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal JSON: %v", err)
	}

	// Print JSON
	fmt.Println(string(jsonData))
	return nil
}

// printTextOutput prints container information in text format
func printTextOutput(containers []container.ContainerInfo) {
	// Create a tabwriter for pretty output
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
	fmt.Fprintln(w, "CONTAINER ID\tNAME\tIMAGE\tSTATE\tTRANSPORT\tPORT")

	// Print container information
	for _, c := range containers {
		// Truncate container ID to first 12 characters (similar to Docker)
		truncatedID := c.ID
		if len(truncatedID) > 12 {
			truncatedID = truncatedID[:12]
		}

		// Get container name from labels
		name := labels.GetContainerName(c.Labels)
		if name == "" {
			name = c.Name // Fallback to container name
		}

		// Get transport type from labels
		transport := labels.GetTransportType(c.Labels)
		if transport == "" {
			transport = "unknown"
		}

		// Get port from labels
		port, err := labels.GetPort(c.Labels)
		if err != nil {
			port = 0
		}

		// Print container information
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%d\n",
			truncatedID,
			name,
			c.Image,
			c.State,
			transport,
			port,
		)
	}

	// Flush the tabwriter
	w.Flush()
}